### ISSUE TITLE
Can you explain the producer/consumer example provided in the documentation?

### ISSUE THREAD

**Issue by leonidumanskiy** [aaugustin/websockets#156]
I struggle to understand the consumer/producer example provided in the documentation:

```
async def handler(websocket, path):
    while True:
        listener_task = asyncio.ensure_future(websocket.recv())
        producer_task = asyncio.ensure_future(producer())  # <- what should producer do?
        done, pending = await asyncio.wait(
            [listener_task, producer_task],
            return_when=asyncio.FIRST_COMPLETED)

        if listener_task in done:
            message = listener_task.result()
            await consumer(message)
        else:
            listener_task.cancel()

        if producer_task in done:
            message = producer_task.result()
            await websocket.send(message)
        else:
            producer_task.cancel()
```

Producer/consumer pattern, from the [wikipedia](https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem):

> The problem describes two processes, the producer and the consumer, who share a common, fixed-size buffer used as a queue. The producer's job is to generate data, put it into the buffer, and start again. At the same time, the consumer is consuming the data (i.e., removing it from the buffer), one piece at a time. 

As far as I understand the consumer/producer pattern, it is used when the producer creates data and puts it into the queue/buffer (i.e. receives the data sent by the client), at the same time consumer takes this data from the queue/buffer and handles it appropriately, one message at a time. The pattern is designed to solve a fixed-buffer problem; when buffer is full, producer must wait for the consumer to handle the data before putting a new data into that buffer.

The provided example doesn't fit into that pattern and doesn't solve any fixed-buffer problem. The consumer is used to handle data received by the client, and the role of producer is not clear for me, in the code it simply takes some data from some method and sends it to the client.

I guess that would make sense if producer and consumer were swapped (producer would receive data from the client, put it into some magic buffer, at the same time consumer would take this data from the buffer, handle it and send the result back to the client), but I am not sure if that would be even right (what if you need to send some data to another client?).

Sorry about writing too much text. I really like the library, but I am trying to write a chat application and I don't really understand what would be the "best practice" way of doing it. I also don't understand why can't we simply send the data to the client by doing ensure_future(websocket.send(data)), that way we don't need to yield the client coroutine.
Thanks.

**Comment by aaugustin** [comment]
The docs don't talk about the "consumer/producer pattern" so I'm not feeling bound by the description in Wikipedia.

The first example is if you're a consumer getting messages over the websocket connection (producer is on the other end of the connection).

The second example is if you're a producer sending messages over the websocket connection (consumer is on the other end of the connection).

If you have better words to describe these roles, I can consider changing them, but until now they haven't caused confusion.

The third example is if you want to do both at the same time, in parallel, which was a common question and isn't entirely obvious to implement with asyncio, so I included an example of how to do that.

However it isn't a "producer/consumer" example, and that's why I didn't describe it as such. It's a producer running alongside a consumer.

----

`ensure_future(websocket.send(data))` works but is less idiomatic and probably more expensive than `yield from websocket.send(data)`. It also makes it more difficult to handle exceptions, you'll just get stack traces in logs instead of being able to catch exceptions.

**Comment by leonidumanskiy** [comment]
Hi aaugustin,
Thanks for your answer! 
Can you please clarify the role of the producer() function in the context of simple chat server?

When building a scalable chat application, this function can take new messages from the MQ or MongoDB and send them over to the client. However, if might be cancelled if it yields (i.e. waiting for the MongoDB query) and the new data comes from the client (producer_task.cancel() will be called).
Or should this function only take data from the sending buffer and yield until new data appeared in this buffer? 

Sorry, the role of the producer() function and what it can or can't implement is still a bit confusing to me.

**Comment by aaugustin** [comment]
If you have a `producer()` coroutine that asynchronously returns the next thing to send through the websocket connection, then the producer example gives you a rough idea of how to do that (minus error handling).

If you don't have such a coroutine and you don't see how to write one, then this example isn't for you. It's just an *example* that works in some use cases. If you want a function to "take new messages from the MQ or MongoDB and send them over to the client", use `websocket.send`.

Regarding the more specific questions, if `producer()` is cancelled, it will be restarted during the next iteration of the loop, so that isn't an issue.

This usually requires an in-process, in-memory bounded queue to get adequate performance. `producer()` usually pops from that queue. Another task fills as new data is published from MQ.
